### ISSUE TITLE
CLI vs FPM/CGI Pool

### ISSUE THREAD

**Issue by ellisonpatterson** [lcache/lcache#77]
### Steps to reproduce
Run the script from FastCGI and then through CLI

### Expected behavior
Pools should match from both instances (although makes sense why it currently does not and not even sure if there's an elegant fix).

### Actual behavior
FastCGI: 192.168.1.1:443
CLI: 599493z9959zc9.20818927:240895499

### Frameworks tested
Which module / plugin were you using?
- Coding Zend Framework integration but direct LCache usage would most likely replicate this issue.

**Comment by davidstrauss** [comment]
Just to start, it's not actually a problem for correctness if FastCGI and the CLI use different pool IDs. The pool ID only create a namespace for the L1 cache; the L2 cache is still shared. Nor does it create any issue for cache consistency; writes from the CLI will still be visible to PHP under FastCGI and vice versa.

Also, in the case of `APCuL1`, it's impossible to share the L1 between the CLI and FastCGI; this is because APCu only stores data in the local process pool.

However, I can see the value in sharing an L1 if it's actually persistent and shareable, like `SQLiteL1`. In that case, you may want to explicitly set the pool for both FastCGI and the CLI so they can share. This would currently cause issues between the CLI and FastCGI because the best state driver for `SQLiteL1` is still backed by APCu.

**Comment by ellisonpatterson** [comment]
Yeah I just used gethostname() which would be different across the servers so it should be all good :)

**Comment by davidstrauss** [comment]
I'm still not sure I'd recommend sharing the L1 and pool between the CLI and FastCGI until there's a SQLite-backed L1 state driver, as sharing L1 state is necessary to share the L1.

**Comment by davidstrauss** [comment]
@ellisonpatterson Also, since you're working on a new consumer for the library, you may want to check out v0.4.0, which I just published. We've made substantial API changes for initializing the cache.

**Comment by ellisonpatterson** [comment]
Alright sounds good 👍 Only issue left is duplicate addresses being written into the db. Could that be related to what I'm doing?

**Comment by davidstrauss** [comment]
> Only issue left is duplicate addresses being written into the db. Could that be related to what I'm doing?

Can you provide an example? There *should* be duplicates in the L2 if multiple events have affected a cache address; those should get pruned eventually under a variety of circumstances.

**Comment by ellisonpatterson** [comment]
![329ac40a07da78d2f20bb53a708b6f2e](https://cloud.githubusercontent.com/assets/3533001/21531692/127d4f58-cd18-11e6-83ce-6596c63a75fd.png)

I'm just seeing hundreds of entries for the same key.

**Comment by davidstrauss** [comment]
> I'm just seeing hundreds of entries for the same key.

Are they being written in the same script? The cleanup currently only prunes items that have `event_id` values preceding the start of the current script; this simplifies the cleanup query a lot.

So, basically, it should work this way:

1. Script run 1: Write to an address. Note the `event_id` for this write.
2. Script run 2: Write to that same address. The destructor for `DatabaseL2` should do a batch deletion of events to addresses affected in this run whose `event_id`s precede this run.
3. The `event_id` from the first run should now be gone.

If this is not happening, then we should open an issue. I'm also considering complicating the desctructor query to get more predictable behavior.

**Comment by ellisonpatterson** [comment]
Currently I have it made so it creates the cache object, creates a new instance of LCache, and does all the fetching and writing there.

```
$cache = array();

$pool = gethostname();
$li = new L1CacheFactory();

$l1 = $li->create('apcu', $pool);

$l2 = new DatabaseL2($db, $tablePrefix, true);

$lcache = new Integrated($l1, $l2);
$lcache->synchronize();

// Writing
foreach ($datas as $data) {
	if (is_null($data)) {
		return (bool) $this->remove($id);
	}

	if (is_object($data)) {
		$data = clone $data;
	}

	$cache[$id] = $data;

	if (!is_numeric($data) || intval($data) !== $data) {
		$data = serialize($data);
	}

	/* Then it creates a new address and calls $lcache->set, finally setting the data */
}
```

That is what it does generally (using 0.4.0 as well).

**Comment by davidstrauss** [comment]
I can't from the code if there's any misbehavior. It would take multiple runs to see the issue. Are you experiencing issue #67?

**Comment by ellisonpatterson** [comment]
Yep I'm experiencing that exact issue.
